painttiming.bs

```diff
@@ -197,21 +199,38 @@ Processing model {#sec-processing-model}
 Reporting paint timing {#sec-reporting-paint-timing}
 --------------------------------------------------------
 
+<h4 dfn>Previously reported paints</h4>
+Every [=Document=] has an associated [=set=] of previously reported paints, initiallized to an empty [=set=].
+
+<h4 dfn export>First Contentful Paint</h4>
+<div algorithm="Compute first contentful paint">
+    When asked to <dfn>compute first contentful paint</dfn> given |document|, perform the following steps:
+    1. If |document|'s [=previously reported paints=] contains <code>"first-contentful-paint"</code>, then return false.
+    1. If |document| contains at least one [=element=] that is both [=paintable=] and [=contentful=], then return true.
+    1. Otherwise, return false.
+</div>
+
 <h4 dfn export>Mark paint timing</h4>
 
 <div algorithm="Mark paint timing">
     When asked to [=mark paint timing=] given a [=Document=] |document| as input, perform the following steps:
     1. If the [=document=]'s [=responsible browsing context=] is not [=paint-timing eligible=], return.
     1. Let |paintTimestamp| be the [=current high resolution time=].
-    1. If this instance of [=update the rendering=] is the [=first paint=] of |document|, then invoke the [[#report-paint-timing]] algorithm with |document|, <code>"first-paint"</code>, and |paintTimestamp|.
+    1. Let |reportedPaints| be the [=previously reported paints=] associated with |document|.
+    1. If |reportedPaints| does not contain <code>"first-paint"</code>, then:
+        1. Invoke the [[#report-paint-timing]] algorithm with |document|, <code>"first-paint"</code>, and |paintTimestamp|.
+
+            NOTE: [=First paint=] excludes the default background paint, but includes non-default background paint.
 
-        NOTE: [=First paint=] excludes the default background paint, but includes non-default background paint.
+        1. [=Append=] <code>"first-paint"</code> to |reportedPaints|.
 
-    1. If this instance of [=update the rendering=] is the [=first contentful paint=] of |document|, then invoke the [[#report-paint-timing]] algorithm with |document|, <code>"first-contentful-paint"</code>, and |paintTimestamp| as arguments.
+    1. If the result of invoking the [=compute first contentful paint=] with |document| returns <code>true</code>, then:
+        1. Invoke the [[#report-paint-timing]] algorithm with |document|, <code>"first-contentful-paint"</code>, and |paintTimestamp| as arguments.
+        1. [=Append=] <code>"first-contentful-paint"</code> to |reportedPaints|.
```

Comment:
This is not needed given the append in 'report paint timing'. Also this is appending to 'reportPaints', not the document's set, so be careful about that. Same for the 'append' above for first paint.